src/buf/uninit_slice.rs

```diff
@@ -0,0 +1,165 @@
+use core::fmt;
+use core::mem::MaybeUninit;
+use core::ops::{
+    Index, IndexMut, Range, RangeFrom, RangeFull, RangeInclusive, RangeTo, RangeToInclusive,
+};
+
+/// Uninititialized byte slice.
+///
+/// Returned by `BufMut::bytes_mut()`, the referenced byte slice may be
+/// uninitialized. The wrapper provides safe access without introducing
+/// undefined behavior.
```

Comment:
It would be good to say what the concrete invariant/protocol for this type is, and in particular how it differs from `MaybeUninit`.

if I understand correctly, the key difference is that the contents may be uninitialized, but the API ensures that once they are initialized, they will never be de-initialized again. Is that correct?

Reply:
Yes, that's correct. Especially it allows constructing `UninitSlice` from already-initialized slice (`&mut [u8]`).